posts/2021/09/array-libraries-interoperability.md

```diff
@@ -0,0 +1,537 @@
+
+The demo shows the implementation of a dummy [`get_namespace`](https://anirudhdagar.ml/array-api-demo/GW_Demo_Array_API.html#get-namespace) method which is the
+first function to be called inside any SciPy method. One can see how it works
+below, simply returning the namespace, which can be used later for calling any
+Array API specified methods.
+
+```python
+def get_namespace(*xs):
+    # `xs` contains one or more arrays, or possibly Python scalars (accepting
+    # those is a matter of taste, but doesn't seem unreasonable).
+    namespaces = {
+        x.__array_namespace__() if hasattr(x, '__array_namespace__') else None for x in xs if not isinstance(x, (bool, int, float, complex))
+    }
+
+    if not namespaces:
+        # one could special-case np.ndarray above or use np.asarray here if
+        # older numpy versions need to be supported.
+        # This can happen when lists are sent as an input, for eg. some
+        # SciPy functions coerce lists into ndarrays internally.
+        raise ValueError("Unrecognized array input")
+
+    if len(namespaces) != 1:
+        raise ValueError(f"Multiple namespaces for array inputs: {namespaces}")
+
+    xp, = namespaces
+    if xp is None:
+        raise ValueError("The input is not a supported array type")
+
+    return xp
+
+
+def csd(x, y, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+	detrend='constant', return_onesided=True, scaling='density', axis=-1,
+	average='mean'):
+    # Array-API
+    xp = get_namespace(x, y)
+
+    # Library Independent Code
+    # Call Array API specified methods
+    some_internal_calculation = xp.some_func(x)
+    ...
+```
+
+This should be possible with the Array API, for the end-user in a
+future release of SciPy and PyTorch. I'd encourage you to read
+more details on the [demo](https://anirudhdagar.ml/array-api-demo/intro.html)
+page itself.
+
+
+### Contributing to PyTorch
+
+Interestingly in the process and effort to make PyTorch
+more [Array API](https://data-apis.org/array-api/latest/) compatible,
+I started contributing to PyTorch by raising PRs and relevant Issues.
+My first PyTorch PR [#62560](https://github.com/pytorch/pytorch/pull/62560)
+started with adding an alias `torch.concat` for `torch.cat` which is
+how concatenation is defined in the [array api spec](https://data-apis.org/array-api/latest/API_specification/manipulation_functions.html?highlight=concat#concat-arrays-axis-0)
+and ended with a bunch of other fixes related to concatenation in PyTorch.
+Later, I worked on improving compatibility for Array API to the `torch.linalg`
+module. See [pytorch/pytorch#63285](https://github.com/pytorch/pytorch/pull/63285)
+and [pytorch/pytorch#63227](https://github.com/pytorch/pytorch/pull/63227).
+
+You can track the progress on PyTorch's github repo with the label
+["module: python array api"](https://github.com/pytorch/pytorch/labels/module%3A%20python%20array%20api) to check out other interesting developments.
+
+ 
+## uarray
```

Comment:
This transition is also not very smooth. What would help a lot is further up explain that for pure Python + NumPy code you use the array API standard, while for compiled code like `fft`, `ndimage`, `special` there is a need for array library & hardware specific implementations, and hence from SciPy we need to be able to access and use those. Which is where uarray comes in.